src/Functions/concat.cpp

```diff
@@ -145,8 +135,18 @@ class ConcatImpl : public IFunction
                 constant_strings[i] = const_col->getValue<String>();
             }
             else
-                throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Illegal column {} of argument of function {}",
-                    column->getName(), getName());
+            {
+                // An arbitrary type argument: converting it to a StringColumn as if `toString` was called
+                ColumnsWithTypeAndName args;
+                args.emplace_back(column, arguments[i].type, "tmp");
+                const ColumnPtr converted_col_ptr = ConvertImplGenericToString<ColumnString>::execute(
```

Comment:
`converted_col_ptr` is a smart pointer (in more detail (but the details don't matter for the sake of the issue): it is similar to a `std::shared_pointer` in the sense that it does ref-counting, but dissimilar in the sense that it has a custom implementation). In l. 142, the refcount is 1.

In l. 144, we cast to `ColumnString *` and store pointers to the two internal structures of `ColumnString` into `data` and `offsets` (l. 147/148).

Next, the block goes out-of-scope (l. 149), the refcount drops to 0 and `converted_col_ptr` is deleted. `data` and `offsets` now contain pointers into nowhere.

To fix this, you need to keep the refcounted-pointer around. A quickfix for me was to add

```cpp
std::vector<ColumnPtr> col_ptrs(num_arguments);
```

after l. 115 and then do

```cpp
col_ptrs[i] = converted_col_ptr;
```

before l. 144.